File: src/main/scala/net/nextlogic/airsim/api/gameplay/chauffeur/ChauffeurDronePlayer.scala
package net.nextlogic.airsim.api.gameplay.chauffeur

import net.nextlogic.airsim.api.gameplay.{AirSimBaseClient, DronePlayer}
import net.nextlogic.airsim.api.utils.{Constants, Vector3r}

case class ChauffeurDronePlayer(vehicle: AirSimBaseClient) extends DronePlayer {
  override def steer(phi: Double): Unit = {
    val dtheta = phi * (vehicle.settings.maxVelocity / Constants.turningRadius)
    theta += dtheta * Constants.timeStepForAngleChange
    println(s"${vehicle.settings.name}: Steering with theta $theta")
  }

  override def evade(relativePos: Vector3r, opponentTheta: Double): Unit = {
    val x = relativePos.x
    val y = relativePos.y

    // control variable, which is effectively turning radius
    val phi = y match {
      case 0 if x < 0 => 1
      case 0 => 0
      case _ => Math.signum(y)
    }

    //System.out.println(relativePos);
    steer(phi)
    move()
  }

  def pursue(relativePos: Vector3r, opponentTheta: Double): Unit = {
    val x = relativePos.x
    val y = relativePos.y
    // control variable, which is effectively turning radius
    println(s"${vehicle.settings.name}: relative distance ($x, $y)")

    val phi = y match {
      case 0 if x < 0 => 1
      case 0 => 0
      case _ => Math.signum(y)
    }

    steer(phi)

    move()
  }

}

File: src/main/java/net/nextlogic/airsim/gameplay/chauffeur/ChauffeurDronePlayer.java
package net.nextlogic.airsim.gameplay.chauffeur;

import net.nextlogic.airsim.Simulator;
import net.nextlogic.airsim.gameplay.DronePlayer;

import java.net.UnknownHostException;

public class ChauffeurDronePlayer extends DronePlayer {
    final static double deltaT = 0.08; // time step for angle change
    protected double minR; // minimum turning radius

    public ChauffeurDronePlayer(String ip, String vehicle, double v, double r) throws UnknownHostException {
        super(ip, vehicle, v);

        minR = r;
    }

    public double getMinR() {
        return minR;
    }

    public void steer(double phi) {
        double dtheta = phi*(maxV/minR);
        theta += dtheta*deltaT;

        System.out.println(vehicle + ": Steering with theta " + theta);
    }

    public void move() {
        Vector3r vel = new Vector3r((float) (maxV*Math.cos(theta)), (float) (maxV*Math.sin(theta)), 0f);
        moveByVelocityZ(vel, new Vector3r(0, 0, Simulator.altitude), dt, DrivetrainType.MaxDegreeOfFreedom,
                new YawMode());

//		moveByAngle(-0.1f, 0f, -5f, (float) theta, dt);

        // updatePositionData();
    }

}
